File: library/src/main/java/io/github/anderscheow/validator/rules/common/NotEqualRule.java
package io.github.anderscheow.validator.rules.common;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.util.Locale;

import io.github.anderscheow.validator.rules.BaseRule;

public class NotEqualRule extends BaseRule {

    private String keyword;

    public NotEqualRule(String keyword) {
        super(String.format(Locale.getDefault(), "Value equal to '%s'", keyword));
        this.keyword = keyword;
    }

    public NotEqualRule(String keyword, @StringRes int errorRes) {
        super(errorRes);
        this.keyword = keyword;
    }

    public NotEqualRule(String keyword, @NonNull String errorMessage) {
        super(errorMessage);
        this.keyword = keyword;
    }

    @Override
    public boolean validate(Object value) {
        if (value instanceof String) {
            return !value.equals(keyword);
        }

        throw new ClassCastException("Required String value");
    }
}

File: library/src/main/java/io/github/anderscheow/validator/rules/common/NotEqualRule.kt
package io.github.anderscheow.validator.rules.common

import androidx.annotation.StringRes
import io.github.anderscheow.validator.rules.Rule

class NotEqualRule : Rule {

    private var keyword: Any

    constructor(keyword: Any, @StringRes errorRes: Int) : super(errorRes) {
        this.keyword = keyword
    }

    constructor(keyword: Any, errorMessage: String) : super(errorMessage) {
        this.keyword = keyword
    }

    override fun validate(value: String?): Boolean {
        if (value == null) {
            throw NullPointerException()
        } else {
            return value != keyword
        }
    }
}

fun notEqualTo(keyword: String, @StringRes errorRes: Int): NotEqualRule =
    NotEqualRule(keyword, errorRes)

fun notEqualTo(keyword: String, errorMessage: String): NotEqualRule =
    NotEqualRule(keyword, errorMessage)